### PR TITLE
✨ Import and re-export data structures from Starlette, used by Request properties, on `fastapi.datastructures`

### LINKED ISSUE #1871: Add imports of Starlette datastructures for type annotating Request
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * Or, I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Description

As a user of FastAPI, I want to type annotate functions that return properties of `fastapi.Request` without needing to import `starlette`, in keeping with the [documentation that states](https://fastapi.tiangolo.com/deployment/#about-starlette):

> You shouldn't pin the version of starlette.

### The solution you would like

Instead of the current state, where Starlette must be a pinned dependency to be able to use type annotations for the properties of `Request`:

```python
from fastapi import Request
from starlette.datastructures import URL  # Can only import URL from starlette; it isn't exposed in FastAPI

def example_get_url_from_request(request: Request) -> URL:
    return request.url
```

Instead, import all necessary properties in `fastapi.requests` such that the example could be updated to:

```python
from fastapi import Request
from fastapi.requests import URL

def example_get_url_from_request(request: Request) -> URL:
    return request.url
```

### PR DESCRIPTION
**Pull request by jamescurtin**
Closes https://github.com/tiangolo/fastapi/issues/1871

### CODE REVIEW

**Comment by SirTelemak** [comment]
Why not import it in fastapi.datastructures for more compatible syntax? Imo it's unintuitive that starlette.datastructures models imported from fastapi.requests.
Moreover in your example:
```python3
def example_get_url_from_request(request: Request) -> URL:
```
there URL is not request, it's response which, again, unintuitive in my opinion.

**Comment by jamescurtin** [comment]
> Why not import it in fastapi.datastructures for more compatible syntax? Imo it's unintuitive that starlette.datastructures models imported from fastapi.requests.

Good point; restructured such that the imports mirror the structure of starlette

File: fastapi/datastructures.py [hunk 1]
```diff
@@ -1,5 +1,11 @@
 from typing import Any, Callable, Iterable, Type
 
+from starlette.datastructures import URL  # noqa: F401
```
Comment (Kludex) [inline fastapi/datastructures.py:3]: I don't know how to create a multiline suggestion, but:
```suggestion
from starlette.datastructures import (
    Address,
    FormData,
    Headers,
    QueryParams,
    State,
    UploadFile as StarletteUploadFile,
    URL
)
```

File: fastapi/__init__.py [hunk 2]
```diff
@@ -1,7 +1,6 @@
 """FastAPI framework, high performance, easy to learn, fast to code, ready for production"""
 
 __version__ = "0.61.0"
-
```
Comment (Kludex) [inline fastapi/__init__.py:4]: Readd this line, please

File: fastapi/types.py [hunk 3]
```diff
@@ -0,0 +1 @@
+from starlette.types import Receive  # noqa: F401
```
Comment (Kludex) [inline fastapi/types.py:1]: Repeated?

**Review by Kludex — COMMENTED** [review]
My suggestions are only focused on the syntax. I'm not sure about the idea of the changes tho. I saw the issue about this, but I'm still not sure.

File: fastapi/__init__.py (hunk 2 above)
Comment (jamescurtin) [reply fastapi/__init__.py:4]: This diff was generated by running `bash scripts/format-imports.sh`. I just confirmed by re-adding it and running the format script; it removes the line again.

File: fastapi/datastructures.py (hunk 1 above)
Comment (jamescurtin) [reply fastapi/datastructures.py:3]: In general I would agree, but the change you suggest is incompatible with the `bash scripts/format-imports.sh` script.

That script invokes `isort fastapi tests docs_src scripts --force-single-line-imports`, which splits each import to a single line, but does not propagate the `noqa` hint to each line. It then invokes `autoflake --remove-all-unused-imports ...` which deletes the imports.

File: fastapi/types.py (hunk 3 above)
Comment (jamescurtin) [reply fastapi/types.py:1]: Yes--thank you! (Though it was actually the one in `fastapi/requests.py` that was repeated)

**Comment by jamescurtin** [comment]
👋  Checking in on this PR: would someone please be able to review? Thank you!

**Comment by tiangolo** [comment]
Great, thanks @jamescurtin for your contribution! :rocket: 

Thanks @Kludex and @SirTelemak for the discussion!

I updated it a bit to keep the current import style that improves typing support in editors (e.g. autoimports).

This will be available later today in FastAPI version `0.68.0` 🔖🎉

**Pull request merged**